smarts/core/utils/math.py

```diff
@@ -218,6 +247,131 @@ def vec_to_radians(v) -> float:
     return (r - 0.5 * math.pi) % (2 * math.pi)  # quad 1
 
 
+def is_close(a, b, rel_tol=1e-09, abs_tol=0.0):
+    return abs(a - b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)
+
+
+def euclidean_distance(p1, p2):
+    dx = p1[0] - p2[0]
+    dy = p1[1] - p2[1]
+    return math.sqrt(dx * dx + dy * dy)
+
+
+def position_at_offset(p1, p2, offset):
+    if is_close(offset, 0.0):  # for pathological cases with dist == 0 and offset == 0
+        return p1
+
+    dist = euclidean_distance(p1, p2)
+
+    if is_close(dist, offset):
+        return p2
+
+    if offset > dist:
+        return None
+
+    return p1[0] + (p2[0] - p1[0]) * (offset / dist), p1[1] + (p2[1] - p1[1]) * (
+        offset / dist
+    )
+
+
+def offset_along_shape(point, shape) -> float:
+    if point not in shape:
+        return polygon_offset_with_minimum_distance_to_point(point, shape)
+    offset = 0
+    for i in range(len(shape) - 1):
+        if shape[i] == point:
+            break
+        offset += euclidean_distance(shape[i], shape[i + 1])
+    return offset
+
+
+def position_at_shape_offset(shape, offset):
+    seen_length = 0
+    curr = shape[0]
+    for next_p in shape[1:]:
+        next_length = euclidean_distance(curr, next_p)
+        if seen_length + next_length > offset:
+            return position_at_offset(curr, next_p, offset - seen_length)
+        seen_length += next_length
+        curr = next_p
+    return shape[-1]
+
+
```

Comment:
These functions are mostly from `sumolib.net.geomhelper.py`. So if we optimize them, we can directly use them in `SumoRoadNetork` class instead of calling from the above file.